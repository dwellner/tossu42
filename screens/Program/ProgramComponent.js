import React from "react";
import PropTypes from "prop-types";
import {
  ScrollView,
  StyleSheet,
  View,
  TouchableOpacity,
  Text
} from "react-native";
import Colors from "../../constants/Colors";
import Styles from "../../constants/Styles";
import { Labels } from "../../constants/Texts";
import ProgramWeekComponent from "./ProgramWeekComponent";
import { LinearGradient } from "expo";
import Logo from "../../components/Logo";

const Header = () => (
  <View
    style={{
      margin: 16,
      alignItems: "stretch"
    }}
  >
    <Text style={{ ...Styles.strongLargeContent, textAlign: "center" }}>
      {Labels.programHeader}
    </Text>
    <Text style={{ ...Styles.defaultContent, textAlign: "center" }}>
      {Labels.programGuide}
    </Text>
  </View>
);

export default class ProgramComponent extends React.Component {
  static propTypes = {
    weekProgram: PropTypes.object.isRequired,
    targetEvent: PropTypes.object.isRequired,
    date: PropTypes.string.isRequired,
    onWeekClicked: PropTypes.func.isRequired
  };

  render() {
    const { date, weekProgram, targetEvent, onWeekClicked } = this.props;

    const weeks = weekProgram.weeks.map((week, index) => {
      const isCurrent = week.days.findIndex(day => day.date === date) >= 0;
      const isPast = week.days[6].date < date;

      const itemStyles = isCurrent
        ? [styles.listItem, styles.currentItem]
        : styles.listItem;
      return (
        <TouchableOpacity
          key={week.days[0].date}
          onPress={() => onWeekClicked(week.days[0].date)}
          style={itemStyles}
        >
          <ProgramWeekComponent
            targetEvent={targetEvent}
            weekNumber={week.weekNumber}
            week={week}
            isPast={isPast}
          />
        </TouchableOpacity>
      );
    });

    return (
      <LinearGradient
        style={{ flex: 1, alignSelf: "stretch" }}
        colors={["#fff", "#eee", "#fff"]}
      >
        <Logo />
        <Header />
        <ScrollView style={{ flex: 1 }}>
          <View style={styles.listContainer}>{weeks}</View>
        </ScrollView>
      </LinearGradient>
    );
  }
}

const styles = StyleSheet.create({
  listContainer: {
    paddingBottom: 30
  },

  listItem: {
    flex: 1,
    marginTop: 16,
    marginTop: 16,
    marginLeft: 24,
    marginRight: 24,
    backgroundColor: "#fff"
  },

  currentItem: {
    borderColor: Colors.tintColor,
    borderWidth: 2
  }
});
