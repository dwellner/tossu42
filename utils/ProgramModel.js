import DateUtils from "./DateUtils";
import ProgramService from "../data/ProgramService";

const getIntensityFactor = runType =>
  ({
    iv: 2.5,
    m: 2,
    r: 2,
    t: 2,
    k: 1.5,
    pi: 1.5,
    pe: 1.5,
    ve: 0.5
  }[runType] || 1);

const getIVTotalDistance = (ivDistance, repeats) => {
  const warmup = 2,
    coolDown = 1,
    recovery = 0.4;
  return warmup + (ivDistance + recovery) * repeats + coolDown;
};

const getDayDistance = day => {
  if (day.type === "iv") return getIVTotalDistance(day.distance, day.repeat);
  if (day.type === "lepo") return 0;
  return day.distance;
};

const getDayIntensity = day =>
  getIntensityFactor(day.type) * getDayDistance(day);

const getWeekDistance = week =>
  week.map(getDayDistance).reduce((a, b) => a + b);

const getIntensity = week => week.map(getDayIntensity).reduce((a, b) => a + b);

const getStretchedProgramStart = (program, programLength, totalLength) => {
  let repeats = [];

  const stretchLengthInWeeks = Math.ceil(totalLength - programLength);
  const repeatBlock = program.weeks.slice(0, program.stretchRules.repeat);
  const numberOfRepeatBlocks = Math.ceil(
    stretchLengthInWeeks / repeatBlock.length
  );
  for (i = 0; i < numberOfRepeatBlocks; i++) {
    repeats = repeats.concat(repeatBlock.map(days => [].concat(days)));
  }
  const excessWeeks = repeats.length - stretchLengthInWeeks;
  return repeats.slice(excessWeeks);
};

const getProgramWeeks = (program, length) => {
  const programLength = program.weeks.length;

  if (length === programLength || program.stretchRules === undefined) {
    return program.weeks;
  }

  if (length > programLength) {
    const stretchedStart = getStretchedProgramStart(
      program,
      programLength,
      length
    );
    return [].concat(stretchedStart, program.weeks);
  }

  // TODO: check total length is within program range, or return mimimum
  return []; // TODO: handle shrinking
};

export default {
  getWeekProgram: (targetDate, { targetTime, name, length }) => {
    if (targetDate === null) return null;
    
    const programs = ProgramService.getPrograms(targetTime, name, length);
    if (programs.length === 0) return null;
    const program = programs[0];

    const programWeeks = getProgramWeeks(program, length);
    const firstDate = DateUtils.nextDate(
      targetDate,
      programWeeks.length * -7 + 1
    );
    const maxWeekDistance = programWeeks
      .map(getWeekDistance)
      .reduce((a, b) => Math.max(a, b), 0);

    const maxIntensity = programWeeks
      .map(getIntensity)
      .reduce((max, w) => Math.max(max, w), 0);

    let i = 0;
    const weeks = programWeeks.map(week => {
      const distance = getWeekDistance(week);
      const intensityLevel = getIntensity(week) / maxIntensity;
      const distanceLevel = distance / maxWeekDistance;
      const days = week.map(day => ({
        ...day,
        date: DateUtils.nextDate(firstDate, i++)
      }));
      return { distance, intensityLevel, distanceLevel, days };
    });
    return { weeks };
  }
};
