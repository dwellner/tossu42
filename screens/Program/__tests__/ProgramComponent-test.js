import "react-native";
import React from "react";
import ProgramComponent from "../ProgramComponent";
import renderer from "react-test-renderer";
import ProgramModel from "../../../utils/ProgramModel";

it("renders correctly", () => {
  const targetEvent = { name: "Foo maraton", date: "2018-10-30" };

  const weekProgram = ProgramModel.getWeekProgram("2018-10-30", {
    targetTime: 240,
    name: "Peruskunto - 4t tavoiteaika",
    length: 24
  });
  const tree = renderer
    .create(
      <ProgramComponent
        weekProgram={weekProgram}
        date={weekProgram.weeks[10].days[3].date}
        targetEvent={targetEvent}
        onWeekClicked={() => {}}
        onGotoToday={() => {}}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
