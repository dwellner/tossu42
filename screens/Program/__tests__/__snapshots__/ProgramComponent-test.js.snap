// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ExponentLinearGradient
  colors={
    Array [
      4294967295,
      4293848814,
      4294967295,
    ]
  }
  locations={null}
  style={
    Object {
      "alignSelf": "stretch",
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#535353",
            "borderRadius": 50,
            "height": 60,
            "justifyContent": "center",
            "width": 60,
          },
          Object {
            "marginLeft": 8,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#535353",
              "fontFamily": "OpenSans",
              "fontSize": 20,
            },
            Object {
              "color": "#fff",
            },
          ]
        }
      >
        T42
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "margin": 16,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#535353",
          "fontFamily": "OpenSans-Bold",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Nykyinen harjoitusohjelma
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#535353",
          "fontFamily": "OpenSans",
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      Alla n√§et ohjelman harjoitusviikot
    </Text>
  </View>
</ExponentLinearGradient>
`;
